rebasebot/bot.py

```diff
@@ -137,32 +138,64 @@ def _add_to_rebase(commit_message, source_repo, tag_policy):
     return tag_policy == "soft"
 
 
-def _do_rebase(gitwd, source, dest, source_repo, tag_policy):
+def _do_rebase(gitwd, source, dest, source_repo, tag_policy, allow_bot_squash, bot_emails):
     logging.info("Performing rebase")
 
     merge_base = gitwd.git.merge_base(f"source/{source.branch}", f"dest/{dest.branch}")
     logging.info("Rebasing from merge base: %s", merge_base)
 
     # Find the list of commits between the merge base and the destination head
     # This should be the list of commits we are carrying on top of the UPSTREAM
-    commits = gitwd.git.log("--reverse", "--pretty=format:%H - %s", "--no-merges",
+    commits = gitwd.git.log("--reverse", "--pretty=format:%H - %s - %aE", "--no-merges",
                             "--ancestry-path", f"{merge_base}..dest/{dest.branch}")
     logging.info("Picking commits: \n%s", commits)
 
+    commits_to_squash = defaultdict(list)
+
     for commit in commits.splitlines():
         # Commit contains the message for logging purposes,
         # trim on the first space to get just the commit sha
-        sha, commit_message = commit.split(" - ")
+        sha, commit_message, committer_email = commit.split(" - ")
 
         if not _add_to_rebase(commit_message, source_repo, tag_policy):
             continue
 
         try:
-            gitwd.git.cherry_pick(f"{sha}", "-Xtheirs")
+            if allow_bot_squash:
+                # There is sometimes a prefix with number and a following + sign
+                # We have to get rid of that part to make sure to get
+                # only the email of the bot
+                is_plus_in_email = committer_email.find("+")
+                email = ""
+                if is_plus_in_email != -1:
+                    _, email = committer_email.split("+")
+                else:
+                    email = committer_email
+                if email in bot_emails:
+                    commits_to_squash[email].append(sha)
+                else:
+                    logging.info(f"cherry-picking {sha}")
+                    gitwd.git.cherry_pick(f"{sha}", "-Xtheirs")
+            else:
+                gitwd.git.cherry_pick(f"{sha}", "-Xtheirs")
         except git.GitCommandError as ex:
             if not _resolve_rebase_conflicts(gitwd):
                 raise RepoException(f"Git rebase failed: {ex}") from ex
 
+    # Here we cherry-pick the bot's commits and then squash them together
+    # We also want the newest bot commit message to represent the squashed commits
+    if allow_bot_squash:
+        for key, value in commits_to_squash.items():
+            try:
+                for sha in value:
+                    gitwd.git.cherry_pick(f"{sha}", "-Xtheirs")
```

Comment:
Can you make sure we log `cherr-picking {sha}` here as we do in other places

Reply:
It might be beneficial to explore the logging capabilities of the `gitpython` library. Maybe just logging settings might be tuned to achieve this across the entire program.